File: account-service/src/main/java/com/wks/servicemarketplace/accountservice/config/ApplicationParameters.java
package com.wks.servicemarketplace.accountservice.config;

import lombok.Builder;
import lombok.NonNull;
import lombok.Value;

@Value
@Builder
public class ApplicationParameters {

    private static final String DEFAULT_SERVER_HOST = "0.0.0.0";
    private static final String DEFAULT_SERVER_PORT = "8080";

    @Builder.Default
    private final String serverHost = DEFAULT_SERVER_HOST;
    @Builder.Default
    private final String serverPort = DEFAULT_SERVER_PORT;
    @NonNull
    private final String jdbcUrl;
    @NonNull
    private final String jdbcPassword;
    @NonNull
    private final String jdbcUsername;
    @NonNull
    private final String amqpHost;
    private final int amqpPort;
    @NonNull
    private final String keycloakKeySetUrl;
}

File: service-provider-service/src/main/kotlin/com/wks/servicemarketplace/serviceproviderservice/config/ApplicationParameters.kt
package com.wks.servicemarketplace.serviceproviderservice.config

import com.ufoscout.properlty.Properlty
import com.ufoscout.properlty.reader.EnvironmentVariablesReader
import com.ufoscout.properlty.reader.SystemPropertiesReader

interface ApplicationParameters {
    val serverHost: String
    val serverPort: Int
    val amqpUri: String
    val jdbcUrl: String
    val jdbcUsername: String
    val jdbcPassword: String
    val clientId: String
    val clientSecret: String
    val authServiceBaseUrl: String
    val outboxIntervalMillis: Long

    companion object {
        const val DEFAULT_OUTBOX_INTERVAL_MILLIS: Long = 300

        fun load() = DefaultApplicationParameters.load()
    }
}

private data class DefaultApplicationParameters(private val provider: Properlty) : ApplicationParameters{
    companion object {
        fun load(): ApplicationParameters {
            val provider = Properlty.builder()
                .caseSensitive(false)
                .add("file:/etc/servicesmarketplace/application.properties", ignoreNotFound = true)
                .add("classpath:application.properties", ignoreNotFound = true)
                .add(EnvironmentVariablesReader().replace("_", "."))
                .add(SystemPropertiesReader())
                .build()
            return DefaultApplicationParameters(provider)
        }
    }

    override val serverHost: String
        get() = provider["server.host", "0.0.0.0"]

    override val serverPort: Int
        get() = provider.getInt("server.port", 8084)

    override val authServiceBaseUrl: String
        get() = provider["authservice.host.url"]!!

    override val clientId: String
        get() = provider["authservice.client.id"]!!

    override val clientSecret: String
        get() = provider["authservice.client.secret"]!!

    override val jdbcUrl: String
        get() = provider["jdbc.url"]!!

    override val jdbcUsername: String
        get() = provider["jdbc.username"]!!

    override val jdbcPassword: String
        get() = provider["jdbc.password"]!!

    override val amqpUri: String
        get() = provider["amqp.uri"]!!

    override val outboxIntervalMillis: Long
        get() = provider.getLong("app.outbox.intervalMillis", 500)

}